src/protocol.rs

```diff
@@ -1,253 +1,516 @@
 use std::net::{Ipv4Addr, Ipv6Addr};
 use std::str::FromStr;
 use std::convert::From;
-use std::io::Cursor;
+use std::io::{Cursor, Write, Result as IoResult};
 use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
 use cid::Cid;
-use integer_encoding::VarIntWriter;
+use integer_encoding::{VarInt, VarIntWriter};
 
 use {Result, Error};
 
-///! # Protocol
+///! # ProtocolId
```

Comment:
why this rename?

Reply:
Do you mean this comment in particular, or the whole enum?

Reply:
I meant the whole enum

Reply:
That's subjective, but I think `ProtocolId` makes it clearer it is not a whole protocol definition for example. 
I admit that in practice it's because in my first draft `AddrComponent` was named `Protocol`.

If you don't like it, I will change it back. I don't have a strong opinion.

Reply:
lets use `Protocol` as it better matches what the go library is doing: https://github.com/multiformats/go-multiaddr/blob/master/protocols.go